### ISSUE TITLE
`carthage archive` should produce a more human-friendly archive structure

### ISSUE THREAD

**Issue by kballard** [Carthage/Carthage#1001]
The current `carthage archive` output produces a zip with a good name, but the actual contents of the zip use the Carthage/Build/$plat/... format. This is not very friendly to anyone who downloads the zipfile manually, because expanding it will give them a folder named Carthage.

Since Carthage itself doesn't actually care about the directory structure of the zip, we should instead change it to the following format:

```
FrameworkName/
    Mac/
        FrameworkName.framework
        FrameworkName.framework.dSYM
    iOS/
        FrameworkName.framework
        FrameworkName.framework.dSYM
        // bcsymbolmaps...
    tvOS/
        ...
```

**Comment by NachoSoto** [comment]
So you suggest removing the "Carthage/Build" levels, right?
I agree, I think `Carthage` should take care of moving the different folders inside of `Carthage/Build` when downloading archives :+1: 

The tricky part is that we'd have to make it compatible with both formats for a while, but it shouldn't be too hard.

**Comment by mdiep** [comment]
I'd be :+1: with this.

We could maybe even take it one step further. I think we already test that the `.framework`s, `.dSYM`s, etc. match the platform we're looking for. Maybe we could completely ignore the hierarchy and naming and only search for files that match the given platform? The format you outlined does make sense to me for `carthage archive`.

> The tricky part is that we'd have to make it compatible with both formats for a while, but it shouldn't be too hard.

Yup. We're doing this with our `Cartfile.resolved` order in #717.

**Comment by ikesyo** [comment]
> I agree, I think `Carthage` should take care of moving the different folders inside of `Carthage/Build` when downloading archives :+1: 

and

> Maybe we could completely ignore the hierarchy and naming and only search for files that match the given platform?

IIUC this should be already done when downloading in the following functions/methods:

- [frameworksInDirectory](https://github.com/Carthage/Carthage/blob/0.11/Source/CarthageKit/Project.swift#L718-L728)
- [copyFrameworkToBuildFolder](https://github.com/Carthage/Carthage/blob/0.11/Source/CarthageKit/Project.swift#L437-L449)
- [copyDSYMToBuildFolderForFramework](https://github.com/Carthage/Carthage/blob/0.11/Source/CarthageKit/Project.swift#L451-L461)
- [copyBCSymbolMapsToBuildFolderForFramework](https://github.com/Carthage/Carthage/blob/0.11/Source/CarthageKit/Project.swift#L463-L475)

**Comment by mdiep** [comment]
I think most of the infrastructure is in place to do things this way, but AFAIK we only currently consider items inside `Carthage/$PLATFORM/`.

**Comment by ikesyo** [comment]
> but AFAIK we only currently consider items inside `Carthage/$PLATFORM/`.

What the code are you referring to? For frameworks, we are using the value in the Info.plist of the frameworks for platform detection. For dSYMs and bcsymbolmap files, we are using frameworks' UUIDs of each architecture. Both don't depend on `Carthage/$PLATFORM/`.

**Comment by kballard** [comment]
> So you suggest removing the "Carthage/Build" levels, right?

Yeah, and adding in a top-level folder named after the framework (with the `.framework` extension stripped).
